blocks/bricks/recurrent/base.py

```diff
@@ -168,28 +168,31 @@ def recurrent_apply(brick, application, application_call,
                     logger.warning("unknown input {}".format(value) +
                                    unknown_scan_input)
 
-            # Ensure that all initial states are available.
-            initial_states = brick.initial_states(batch_size, as_dict=True,
-                                                  *args, **kwargs)
-            for state_name in application.states:
-                dim = brick.get_dim(state_name)
-                if state_name in kwargs:
-                    if isinstance(kwargs[state_name], NdarrayInitialization):
-                        kwargs[state_name] = tensor.alloc(
-                            kwargs[state_name].generate(brick.rng, (1, dim)),
-                            batch_size, dim)
-                    elif isinstance(kwargs[state_name], Application):
-                        kwargs[state_name] = (
-                            kwargs[state_name](state_name, batch_size,
-                                               *args, **kwargs))
-                else:
-                    try:
-                        kwargs[state_name] = initial_states[state_name]
-                    except KeyError:
-                        raise KeyError(
-                            "no initial state for '{}' of the brick {}".format(
-                                state_name, brick.name))
-            states_given = dict_subset(kwargs, application.states)
+            if len(application.states):
+                # Ensure that all initial states are available.
+                initial_states = brick.initial_states(batch_size, as_dict=True,
+                                                      *args, **kwargs)
+                for state_name in application.states:
```

Comment:
It seems like the code starting from this line can be moved out of the `if` clause, and the `else` part is not really necessary. Right now we pay a high price of having an additional level of indentation for this new feature, and it would be great to keep the complexity of the code down.